### PR TITLE
Feature: Implement `unsafeBody` on `Request`

### LINKED ISSUE #890: Add `unsafeBody` to `Request`
**Is your feature request related to a problem? Please describe.**
I want to be able to have a much lower-level access to the netty channel to implement custom protocols and message decoders.

**Describe the solution you'd like**
```scala

// Wrapper for HttpContent
final case class UnsafeContent(content: HttpContent) extends AnyVal {
  def isLast: Boolean = content.isInstanceOf[LastContent]
}

// Wrapper for ChannelHandlerContext
final case class UnsafeChannel(ctx: ChannelHandlerContent) extends AnyVal {
  def write(content: Content): Unit
  def writeAndFlush(content: Content): Unit
  def read(): Unit
  def close(): Unit
  def flush(): Unit
}


trait Request {
  // ... 
  // ...
  def unsafeBody(msg: (UnsafeChannel, UnsafeContent) => Unit): Unit
  // ...
  // ...
}
```

**Additional context**
- This will be later used as a primitive to implement `multipart` and other `streaming` solutions.
- This will also assume that the channel is not automatically reading.

### PR DESCRIPTION
**Pull request by amitksingh1490**
fixes #890
- [x] Cleanup

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/service/RequestBodyHandler.scala [hunk 1]
```diff
@@ -2,16 +2,23 @@ package zhttp.service
...
+  msgCallback: (UnsafeChannel, UnsafeContent) => Unit,
+  config: Server.Config[R, Throwable],
+) extends SimpleChannelInboundHandler[Any](true) {
 
   override def channelRead0(ctx: ChannelHandlerContext, msg: Any): Unit = {
 
     if (msg.isInstanceOf[HttpContent]) {
-      msgCallback(UnsafeChannel(ctx), UnsafeContent(msg.asInstanceOf[HttpContent]))
+      val httpContent = msg.asInstanceOf[HttpContent]
+      if (httpContent.content().readableBytes() > config.maxRequestSize)
```
Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/service/RequestBodyHandler.scala]: condition is not correct, config should be passed to the unsafe body and should be handled there

File: example/src/main/scala/example/HelloWorld.scala [hunk 2]
```diff
@@ -6,9 +6,11 @@ import zio._
 object HelloWorld extends App {
 
   // Create HTTP route
-  val app: HttpApp[Any, Nothing] = Http.collect[Request] {
-    case Method.GET -> !! / "text" => Response.text("Hello World!")
-    case Method.GET -> !! / "json" => Response.json("""{"greetings": "Hello World!"}""")
+  val app = Http.collectZIO[Request] { case req =>
+    req.getBodyAsString.map { content =>
```
Comment (amitksingh1490) [inline example/src/main/scala/example/HelloWorld.scala]: unwanted change

File: zio-http/src/main/scala/zhttp/http/Request.scala [hunk 3]
```diff
@@ -96,18 +148,23 @@ object Request {
     url: URL = URL.root,
     headers: Headers = Headers.empty,
     remoteAddress: Option[InetAddress] = None,
-    data: HttpData = HttpData.Empty,
+    @unused data: HttpData = HttpData.Empty,
```
Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/http/Request.scala]: smartly handle body based on the type

File: zio-http/src/main/scala/zhttp/service/Handler.scala [hunk 4]
```diff
@@ -15,41 +15,77 @@ private[zhttp] final case class Handler[R](
...
+  override def channelRead0(ctx: Ctx, msg: Any): Unit = {
     implicit val iCtx: ChannelHandlerContext = ctx
-    unsafeRun(
-      jReq,
-      app,
-      new Request {
-        override def method: Method = Method.fromHttpMethod(jReq.method())
+    msg match {
+      case jReq: HttpRequest    =>
+        val hasLargeContent =
```
Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/service/Handler.scala]: this is not correct as if you read body as stream this is not required

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 5]
```diff
@@ -126,11 +141,33 @@ object ServerSpec extends HttpRunnableSpec {
...
+        val app: Http[Any, Throwable, Request, Response] = Http.collectZIO[Request] { case req =>
+          req.getBodyAsString.map { content =>
+            Response.text(content)
+          }
+        }
+        val res = contentM.flatMap(c => app.deploy.getStatus.run(!!, Method.POST, c.map(_.value).getOrElse("")))
+        assertM(res)(equalTo(Status.REQUEST_ENTITY_TOO_LARGE))
+      } +
+      testM("POST Request.getBodyAsStream") {
+        val app: Http[Any, Throwable, Request, Response] = Http.collectZIO[Request] { case req =>
```
Comment (amitksingh1490) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: use generator to create a big content

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 6]
```diff
@@ -174,8 +211,10 @@ object ServerSpec extends HttpRunnableSpec {
       } +
       testM("echo streaming") {
         val res = Http
-          .collectHttp[Request] { case req =>
-            Http.fromStream(ZStream.fromEffect(req.getBody).flattenChunks)
+          .collectZIO[Request] { case req =>
```
Comment (amitksingh1490) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: use req.getBodyAsStream

File: zio-http/src/main/scala/zhttp/http/Request.scala [hunk 7]
```diff
@@ -22,21 +30,67 @@ trait Request extends HeaderExtension[Request] { self =>
...
 
+  def unsafeBody(msg: (UnsafeChannel, UnsafeContent) => Unit): Unit
+
   /**
    * Decodes the content of request as a Chunk of Bytes
    */
-  def getBody: Task[Chunk[Byte]] =
-    getBodyAsByteBuf.flatMap(buf => Task(Chunk.fromArray(ByteBufUtil.getBytes(buf))))
+  def getBody: Queue[ByteBuf] = ???
+//    for {
```
Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/http/Request.scala]: I think this is not needed as we have a getBodyAsStream anyways, ZStream.effectAsyc uses a queue internally

File: zio-http/src/main/scala/zhttp/http/Request.scala [hunk 8]
```diff
@@ -127,11 +184,16 @@ object Request {
...
-    override def getHeaders: Headers                            = req.getHeaders
-    override def method: Method                                 = req.method
-    override def remoteAddress: Option[InetAddress]             = req.remoteAddress
-    override def url: URL                                       = req.url
-    override private[zhttp] def getBodyAsByteBuf: Task[ByteBuf] = req.getBodyAsByteBuf
+    override def getHeaders: Headers                = req.getHeaders
+    override def method: Method                     = req.method
+    override def remoteAddress: Option[InetAddress] = req.remoteAddress
+    override def url: URL                           = req.url
+    override def unsafeBody(
```
Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/http/Request.scala]: lets keep this private to zhttp for now

File: zio-http/src/main/scala/zhttp/service/server/ServerChannelInitializer.scala [hunk 9]
```diff
@@ -47,7 +47,7 @@ final case class ServerChannelInitializer[R](
 
     // ObjectAggregator
     // Always add ObjectAggregator
-    pipeline.addLast(OBJECT_AGGREGATOR, new HttpObjectAggregator(cfg.maxRequestSize))
+    // pipeline.addLast(OBJECT_AGGREGATOR, new HttpObjectAggregator(cfg.maxRequestSize))
```
Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/service/server/ServerChannelInitializer.scala]: delete commented code, check what other handling is done by Object Aggregator and handle it in getBodyAsString

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 10]
```diff
@@ -174,8 +211,8 @@ object ServerSpec extends HttpRunnableSpec {
...
```
Comment (amitksingh1490) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: use collectHttp

File: zio-http/src/main/scala/zhttp/http/Request.scala (hunk 7 above)
Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/http/Request.scala]: `getBody` is required and it should return a `Chunk[Byte]`

File: zio-http/src/main/scala/zhttp/http/Request.scala [hunk 11]
```diff
@@ -22,21 +30,67 @@ trait Request extends HeaderExtension[Request] { self =>
...
+          cb(IO.succeed(Chunk(msg.content.content())))
+          if (msg.isLast) {
+            ch.ctx.pipeline().remove(HTTP_CONTENT_HANDLER)
+            cb(IO.fail(None))
+          } else {
+            ch.read()
+          }
+
+        }),
+      )
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Request.scala:89]: We can add a `getBodyChunk` which would look like this — 

```scala
def getBodyChunk: ZIO[Any, Option[Throwable], Chunk[Byte]]` = ???
```

This method will internally use `unsafeBody` and read the channel, one message at a time.

File: zio-http/src/main/scala/zhttp/http/Request.scala (hunk 3 above)
Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/http/Request.scala]: I am guessing this still needs to be implemented?

File: zio-http/src/main/scala/zhttp/http/Request.scala [hunk 12]
```diff
@@ -22,21 +30,67 @@ trait Request extends HeaderExtension[Request] { self =>
...
+      override def unsafeBody(
+        msg: (
+          UnsafeChannel,
+          UnsafeContent,
+        ) => Unit,
+      ): Unit = self.unsafeBody(msg)
     }
   }
 
+  def unsafeBody(msg: (UnsafeChannel, UnsafeContent) => Unit): Unit
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Request.scala]: Let's keep it as package private.

**Pull request closed without merging**